forc-pkg/src/pkg.rs

```diff
@@ -2369,14 +2402,66 @@ pub fn build_with_options(build_options: BuildOpts) -> Result<Built> {
     };
 
     let manifest_file = ManifestFile::from_dir(&this_dir)?;
+    let member_manifests = manifest_file.member_manifests()?;
+    let lock_path = manifest_file.lock_path()?;
+    let build_plan = BuildPlan::from_lock_and_manifests(
+        &lock_path,
+        &member_manifests,
+        build_options.pkg.locked,
+        build_options.pkg.offline,
+    )?;
+    let graph = build_plan.graph();
+    let manifest_map = build_plan.manifest_map();
+    let build_profiles: HashMap<String, BuildProfile> = build_plan.build_profiles().collect();
+    // Get the selected build profile using build options
+    let (profile_name, build_profile) = build_profile_from_opts(&build_profiles, &build_options)?;
+
+    // If this is a workspace we want to have all members in the output.
+    let outputs = match &manifest_file {
+        ManifestFile::Package(pkg_manifest) => std::iter::once(
+            build_plan
+                .find_member_index(&pkg_manifest.project.name)
+                .ok_or_else(|| anyhow!("Cannot found project node in the graph"))?,
+        )
+        .collect(),
+        ManifestFile::Workspace(_) => build_plan.member_nodes().collect(),
+    };
+    // Build it!
+    let mut built_workspaces = HashMap::new();
+    let built_packages_with_source_map = build(&build_plan, &build_profile, &outputs)?;
+    let output_dir = build_options
+        .clone()
+        .pkg
+        .output_directory
+        .map(PathBuf::from);
+    for (node_ix, (built_package, source_map)) in built_packages_with_source_map
+        .iter()
+        .filter(|(node_ix, _)| outputs.contains(node_ix))
+    {
+        let pinned = &graph[*node_ix];
+        let pkg_manifest = manifest_map
+            .get(&pinned.id())
+            .ok_or_else(|| anyhow!("Couldn't find member manifest for {}", pinned.name))?;
+        let output_dir = output_dir
+            .clone()
+            .unwrap_or_else(|| default_output_directory(pkg_manifest.dir()).join(&profile_name));
+        built_package.output_artifacts(
+            source_map,
+            build_options.clone(),
+            pkg_manifest,
+            &output_dir,
+        )?;
+        built_workspaces.insert(pinned.name.clone(), built_package.clone());
+    }
+
     match manifest_file {
-        ManifestFile::Package(package_manifest) => {
-            let built_package = build_package_with_options(&package_manifest, build_options)?;
-            Ok(Built::Package(Box::new(built_package)))
-        }
-        ManifestFile::Workspace(_) => {
-            bail!("Workspace building is not supported")
+        ManifestFile::Package(pkg_manifest) => {
+            let built_pkg = built_workspaces
+                .get(&pkg_manifest.project.name)
+                .ok_or_else(|| anyhow!("Couldn't find the built package"))?;
```

Comment:
In this case it might be better to use use `expect` as not having this package within `built_workspace` would be an error in our logic and should never occur.
```suggestion
                .expect("package didn't exist in workspace");
```